Gems/ROS2/Code/Source/VehicleDynamics/DriveModels/AckermannDriveModel.cpp

```diff
@@ -0,0 +1,168 @@
+    {
+        m_driveWheelsData.clear();
+        m_steeringData.clear();
+        m_vehicleConfiguration = vehicleConfig;
+        m_speedPid.InitializePid();
+        m_steeringPid.InitializePid();
+    }
+
+    void AckermannDriveModel::ApplyInputState(const VehicleInputsState& inputs, uint64_t deltaTimeNs)
+    {
+        if (m_driveWheelsData.empty())
+        {
+            m_driveWheelsData = VehicleDynamics::Utilities::GetAllDriveWheelsData(m_vehicleConfiguration);
+        }
+
+        if (m_steeringData.empty())
+        {
+            m_steeringData = VehicleDynamics::Utilities::GetAllSteeringEntitiesData(m_vehicleConfiguration);
+        }
+
+        ApplySteering(inputs.m_steering.GetValue(), deltaTimeNs);
+        ApplySpeed(inputs.m_speed.GetValue(), deltaTimeNs);
+    }
+
+    void AckermannDriveModel::ApplyWheelSteering(SteeringDynamicsData& wheelData, float steering, double deltaTimeNs)
+    {
+        const double deltaTimeSec = double(deltaTimeNs) / 1e9;
+
+        const auto& steeringEntity = wheelData.m_steeringEntity;
+        AZ::Vector3 currentSteeringElementRotation;
+        AZ::TransformBus::EventResult(currentSteeringElementRotation, steeringEntity, &AZ::TransformBus::Events::GetLocalRotation);
+        const float currentSteeringAngle = currentSteeringElementRotation.Dot(wheelData.m_turnAxis);
+        const double pidCommand = m_steeringPid.ComputeCommand(steering - currentSteeringAngle, deltaTimeNs);
+        if (AZ::IsClose(pidCommand, 0.0)) // TODO - use the third argument with some reasonable value which means "close enough"
+        {
+            return;
+        }
+
+        const float torque = pidCommand * deltaTimeSec;
+        AZ::Transform steeringElementTransform;
+        AZ::TransformBus::EventResult(steeringElementTransform, steeringEntity, &AZ::TransformBus::Events::GetWorldTM);
+        const auto transformedTorqueVector = steeringElementTransform.TransformVector(wheelData.m_turnAxis * torque);
+        Physics::RigidBodyRequestBus::Event(steeringEntity, &Physics::RigidBodyRequests::ApplyAngularImpulse, transformedTorqueVector);
+    }
+
+    // TODO - speed and steering handling is quite similar, possible to refactor?
+    void AckermannDriveModel::ApplySteering(float steering, uint64_t deltaTimeNs)
+    {
+        if (m_disabled)
+        {
+            return;
+        }
+        if (m_steeringData.empty())
+        {
+            AZ_Warning("ApplySteering", false, "Cannot apply steering since no steering elements are defined in the model");
+            return;
+        }
+
+        auto innerSteering = atan(
+            (m_vehicleConfiguration.m_wheelbase * tan(steering)) /
+            (m_vehicleConfiguration.m_wheelbase - 0.5 * m_vehicleConfiguration.m_track * tan(steering)));
+        auto outerSteering = atan(
+            (m_vehicleConfiguration.m_wheelbase * tan(steering)) /
+            (m_vehicleConfiguration.m_wheelbase + 0.5 * m_vehicleConfiguration.m_track * tan(steering)));
```

Comment:
Isn't atan2 the safer function to call, when the cosine is 90 degrees or 270 degrees?
That result in divide by 0 error.